pinot-common/src/main/java/org/apache/pinot/common/lineage/SegmentLineage.java

```diff
@@ -0,0 +1,153 @@
+    return _tableNameWithType;
+  }
+
+  /**
+   * Add lineage entry to the segment lineage metadata
+   * @param lineageEntry a lineage entry
+   * @return the id for the input lineage entry for the access
+   */
+  public String addLineageEntry(LineageEntry lineageEntry) {
+    String lineageId = generateLineageId();
+    _lineageEntries.put(lineageId, lineageEntry);
+    return lineageId;
+  }
+
+  /**
+   * Retrieve lineage entry
+   * @param lineageEntryId the id for the lineage entry
+   * @return the lineage entry for the given lineage entry id
+   */
+  public LineageEntry getLineageEntry(String lineageEntryId) {
+    return _lineageEntries.get(lineageEntryId);
+  }
+
+  /**
+   * Delete lineage entry
+   * @param lineageEntryId the id for the lineage entry
+   */
+  public void deleteLineageEntry(String lineageEntryId) {
+    _lineageEntries.remove(lineageEntryId);
+  }
+
+  /**
+   * Convert ZNRecord to segment lineage
+   * @param record ZNRecord representation of the segment lineage
+   * @return the segment lineage object
+   */
+  public static SegmentLineage fromZNRecord(ZNRecord record) {
+    String tableNameWithType = record.getId();
+    Map<String, LineageEntry> lineageEntries = new HashMap<>();
+    Map<String, List<String>> listFields = record.getListFields();
+    for (Map.Entry<String, List<String>> listField : listFields.entrySet()) {
+      String lineageId = listField.getKey();
+      List<String> value = listField.getValue();
+      Preconditions.checkState(value.size() == 4);
+      List<String> segmentsFrom = Arrays.asList(value.get(0).split(COMMA_SEPARATOR));
+      List<String> segmentsTo = Arrays.asList(value.get(1).split(COMMA_SEPARATOR));
+      LineageEntryState state = LineageEntryState.valueOf(value.get(2));
+      long timestamp = Long.parseLong(value.get(3));
+      lineageEntries.put(lineageId, new LineageEntry(segmentsFrom, segmentsTo, state, timestamp));
+    }
+    return new SegmentLineage(tableNameWithType, lineageEntries);
+  }
+
+  /**
+   * Convert the segment lineage object to the ZNRecord
+   * @return ZNRecord representation of the segment lineage
+   */
+  public ZNRecord toZNRecord() {
+    ZNRecord znRecord = new ZNRecord(_tableNameWithType);
+    for (Map.Entry<String, LineageEntry> entry : _lineageEntries.entrySet()) {
+      LineageEntry lineageEntry = entry.getValue();
+      String segmentsFrom = String.join(",", lineageEntry.getSegmentsFrom());
+      String segmentsTo = String.join(",", lineageEntry.getSegmentsTo());
+      String state = lineageEntry.getState().toString();
+      String timestamp = Long.toString(lineageEntry.getTimestamp());
+      List<String> listEntry = Arrays.asList(segmentsFrom, segmentsTo, state, timestamp);
+      znRecord.setListField(entry.getKey(), listEntry);
+    }
+    return znRecord;
+  }
+
+  private String generateLineageId() {
+    return UUID.randomUUID().toString();
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
I don't think `equals()` and `hashCode()` is required for this class.

Reply:
removed